js/ui/number_box/number_box.mask.js

```diff
@@ -275,25 +270,51 @@ var NumberBoxMask = NumberBoxBase.inherit({
             return null;
         }
 
-        parsed = isNaN(parsed) ? undefined : parsed;
+        if(isNaN(parsed)) {
+            return undefined;
+        }
 
-        return this._isPercentFormat() ? (parsed && parsed / 100) : parsed;
-    },
+        var precision = this._getMaxPrecision(format, parsed),
+            pow = Math.pow(10, precision);
 
-    _cleanText: function(text) {
-        var decimalSeparator = number.getDecimalSeparator(),
-            clearRegExp = new RegExp("[^0-9" + decimalSeparator + "]", "g"),
-            cleanedText = text.replace(clearRegExp, "").replace(decimalSeparator, ".");
+        parsed = Math.round((parsed === null ? this._parsedValue : parsed) * pow) / pow;
 
-        return cleanedText;
+        return this._isPercentFormat() ? (parsed && parsed / 100) : parsed;
     },
 
     _isValueIncomplete: function(text) {
         if(!this._useMaskBehavior()) {
             return this.callBase(text);
         }
 
-        return this._cleanText(text).match(/\.0*$/);
+        var formatParts = this._getFormatPattern().split(";")[0].split("."),
+            isFloatPartAllowed = formatParts.length === 2;
+
+        if(!isFloatPartAllowed) {
+            return false;
+        }
+
+        var clearedText = this._removeStubInText(text),
+            decimalSeparator = number.getDecimalSeparator(),
+            decimalSeparatorIndex = clearedText.indexOf(decimalSeparator),
+            separatorIsFirst = decimalSeparatorIndex === 0,
+            lastChar = clearedText.charAt(clearedText.length - 1),
+            onlyOneSeparatorExists = decimalSeparatorIndex === clearedText.length - 1;
+
+        if(separatorIsFirst) {
```

Comment:
Can we combine conditions?
```
if(separatorIsFirst || lastChar !== "0")

Reply:
No, we can't. lastChar may be not equal to zero but equal to decimalSeparator. In this case the method should return true:

```
if(lastChar === decimalSeparator && onlyOneSeparatorExists) {
   return true;
}
```